### PR TITLE
Native fiber

### PR DESCRIPTION
**Pull request by ioquatix**
This is an implementation of fibers using hand-optimised assembly.

There are implementations for amd64, win32, win64, arm32, arm64, and I will add i386 soon too.

### CODE REVIEW

**Comment by ioquatix** [comment]
@nobu I don't expect to merge this yet, but I would like to continue to work on it.

**Comment by ioquatix** [comment]
@nobu I believe this is ready for your review/testing.

File: benchmark/vm2_fiber_chain.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env ruby
+
+require 'benchmark'
```
Comment (k0kubun) [inline benchmark/vm2_fiber_chain.rb]: Any scripts in `benchmark/*.rb` or `benchmark/*.yml` are [expected to be runnable as benchmark-driver's input](https://github.com/ruby/ruby/blob/df58e74e0254c329e8c4fe748a266910070a71a1/common.mk#L1118). Could you change this script for benchmark-driver's format (at least that wouldn't `require 'benchmark'`) or create sub-directory for this under `benchmark/`?

Comment (ioquatix) [reply benchmark/vm2_fiber_chain.rb]: Thanks, I will fix it.

**Comment by ioquatix** [comment]
Looks like build failure on Linux. I will review.

File: benchmark/vm2_fiber_chain.rb (hunk 1 above)
Comment (ioquatix) [reply benchmark/vm2_fiber_chain.rb]: So, I'm happy to change this.

However, right now I'm using `time = Benchmark.realtime do ...` to measure the time of specific parts of the script. I can remove this, or is there a better alternative?

**Comment by ioquatix** [comment]
I am going to see if I can get 32-bit linux passing too. It seems like a very minor target, but it's probably good to support it. As long as it's similar to win32 implementation, it should be straight forward.

**Comment by ioquatix** [comment]
It seems like windows build might be broken. It's selecting  a coroutine implementation (amd64), but I haven't really tested it much. Looks like we might need to fix this so that it uses the existing code path.

```
compiling ../ruby/cont.c
880../ruby/cont.c: In function 'cont_free':
881../ruby/cont.c:416:6: warning: implicit declaration of function 'munmap' [-Wimplicit-function-declaration]
882      munmap((void*)fib->ss_sp, fib->ss_size);
883      ^~~~~~
884../ruby/cont.c: In function 'fiber_initialize_machine_stack_context':
885../ruby/cont.c:909:11: warning: implicit declaration of function 'fiber_machine_stack_alloc' [-Wimplicit-function-declaration]
886     ptr = fiber_machine_stack_alloc(size);
887           ^~~~~~~~~~~~~~~~~~~~~~~~~
888../ruby/cont.c:909:9: warning: assignment to 'char *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
889     ptr = fiber_machine_stack_alloc(size);
890         ^
891../ruby/cont.c:912:41: warning: passing argument 2 of 'coroutine_initialize' from incompatible pointer type [-Wincompatible-pointer-types]
892     coroutine_initialize(&fib->context, fiber_entry, ptr+size, size);
893                                         ^~~~~~~~~~~
894In file included from ../ruby/cont.c:46:
895../ruby/coroutine/amd64/Context.h:30:18: note: expected 'coroutine_start' {aka '__attribute__((noreturn)) void (*)(struct <anonymous> *, struct <anonymous> *)'} but argument is of type '__attribute__((noreturn)) void (*)(void *)'
896  coroutine_start start,
897  ~~~~~~~~~~~~~~~~^~~~~
```

**Comment by ioquatix** [comment]
Here is a brief summary of native implementations now working:

- Darwin (64-bit)
- Linux (32 & 64 bit)
- Windows (64-bit MinGW)

I don't know how to make other Windows builds compile the right files.

Other platforms use existing code path.

**Comment by ioquatix** [comment]
@MSP-Greg I am tracking issues here.

https://github.com/ruby/ruby/commit/c745c56a08b78762b7091fa0974232298a17ef08#commitcomment-31373680

It seems odd to me that MSYS build is failing while win32/win64 build is passing. I wonder what configure is doing.

**Comment by ioquatix** [comment]
```
[00:04:38] checking native coroutine implementation for x64-mingw32... win64
```

**Comment by ioquatix** [comment]
Okay so if you skip those tests everything works fine? So there is still a bug we are just avoiding it right?

**Comment by ioquatix** [comment]
The odd thing is that this works fine in non-MSYS build. It's using almost identical assembly.

**Comment by MSP-Greg** [comment]
Sort of.  It seems to be intermittent, which has been an issue in the Windows builds here.  The SEGV's in test_enumerable.rb have been silent, but I have one from test_enum.rb.  Attached.

[test_all_err.log](https://github.com/ruby/ruby/files/2602003/test_all_err.log)

**Comment by ioquatix** [comment]
Do you have some time to help me reproduce this issue on my windows box? I have windows set up with msys2 using ruby installer and also visual studio 2017. I don't know how to reproduce this error on my setup. I assume I just fetch code using git and then build like usual? Are there build instructions for windows? What do you recommend?

**Comment by ioquatix** [comment]
I wonder if MSYS has a slightly different ABI.

**Comment by ioquatix** [comment]
It's crashing in `_chkstk`. That's an indication the `gs` register might not be set correctly.

**Comment by MSP-Greg** [comment]
> Do you have some time to help me reproduce this issue on my windows box

Yes.  Do you want to repo the test-all errors?

> It's crashing in _chkstk.

Is that from the SEGV log?  Just trying to follow...

**Comment by ioquatix** [comment]
So, while this WAS passing when I tested it, if it's an intermittent problem I might have missed it.

Because I wanted to get this merged, I left windows stack to use malloc. It should be fine, but it's not secure, I'm planning to fix it shortly.

That being said:

```
C:\windows\SYSTEM32\ntdll.dll(_C_specific_handler+0x96) [0x00007ff9be3d1e26]
C:\windows\SYSTEM32\ntdll.dll(_chkstk+0x9d) [0x00007ff9be3e349d]
C:\windows\SYSTEM32\ntdll.dll(RtlRaiseException+0xf67) [0x00007ff9be3a48d7]
C:\windows\SYSTEM32\ntdll.dll(KiUserExceptionDispatcher+0x3a) [0x00007ff9be3e262a]
C:\projects\ruby-loco\install\bin\x64-msvcrt-ruby260.dll(rb_vm_bh_to_procval+0x33) [0x000000006a632ad3]
C:\projects\ruby-loco\install\bin\x64-msvcrt-ruby260.dll(rb_vm_bh_to_procval+0xac) [0x000000006a632b4c]
C:\projects\ruby-loco\install\bin\x64-msvcrt-ruby260.dll(rb_vm_bh_to_procval+0x1dde) [0x000000006a63487e]
```

You can see `_chkstk`. It's a function which reminded me of some other implementations of coroutines. It walks the stack pointer to find stack bounds... I don't know if it's the problem actually, but I know stuff needs to be set up correctly in order for it to work.

This is the information it was documented as required:

https://github.com/ruby/ruby/blob/b755c26146547995b6b4a7c8a05bbae770af0d6b/coroutine/win64/Context.h#L54-L57

That being said, maybe it also need `gs:0` to be set correctly? I don't know. The same code in the `.h` file is used for both MSYS and VC builds... so it's confusing me why it would make any difference. I can only imagine there is a hidden bug in both builds.

**Comment by ioquatix** [comment]
Here are some examples of other similar issues (that I was reminded of when I saw `_chkstk` in the trace):

https://svn.boost.org/trac10/ticket/12592

https://probablydance.com/2013/02/20/handmade-coroutines-for-windows/

**Comment by ioquatix** [comment]
> The bottom stack address is required to be aligned at a page boundary (4KB) on 64bit Windows. protected_fixedsize_stack uses

```
void * vp = ::VirtualAlloc( 0, size__, MEM_COMMIT, PAGE_READWRITE);
```

to allocate its memory, which does have this guarantee (the function allocates on a memory page level). fixedsize_stack however simply uses

```
void * vp = std::malloc( size_);
```

which has no such guarantee.

So, that might be the problem.

**Comment by MSP-Greg** [comment]
Thanks for explanation re `_chkstk`.  Remember, I'm not a c type.

I'm happy to try a patch.  The intermittent issue makes it a bit messier.  At some point hopefully the two skips can be removed...

**Comment by ioquatix** [comment]
I am just pushing it now, using VirtualAlloc... let's see how it goes.

**Comment by ioquatix** [comment]
@MSP-Greg You can see latest change, it's one line. Do you want to try it locally too?

**Comment by ioquatix** [comment]
> Yes. Do you want to repo the test-all errors?

Sure, do you have google hangouts?

**Comment by MSP-Greg** [comment]
I've already added it to ruby-loco as a patch, as soon as the current build is done, I'll start it.

**Comment by ioquatix** [comment]
Here is how we should be allocating stack on Windows.

**Comment by ioquatix** [comment]
@MSP-Greg Sorry, there are some more changes, according to documentation, to correctly allocate guard page.

**Comment by MSP-Greg** [comment]
> there are some more changes

Thanks.  I was just about to push the first patch.  That build passed test-all, but spec froze.  When ` 
Fix missing semi-colon` gets a few minutes into the build, I'll push all the patches to ruby-loco.

> Sure, do you have google hangouts?

I haven't done much of that, as I haven't replaced the (nice) desktop I had.  I'm now on a rather light notebook.  I don't even have VS installed.

Re the ruby-loco builds, every build saves an artifact, so you can dl them and run tests with runner.rb.  Building is a bit more involved, but not that complicated.

**Comment by ioquatix** [comment]
It looks like everything is compiling now.

**Comment by MSP-Greg** [comment]
> It looks like everything is compiling now.

I'm building at https://ci.appveyor.com/project/MSP-Greg/ruby-loco/builds/20448476

**Comment by ioquatix** [comment]
Looks like still got problems :(

**Comment by ioquatix** [comment]
How did you get stack trace before, did you run it locally?

**Comment by ioquatix** [comment]
Oh, I know the problem... I'm still using free on virtualalloc memory lol

**Comment by MSP-Greg** [comment]
> How did you get stack trace before, did you run it locally?

It was collected by the ruby-loco build.  I hate long CI logs, so everything is saved to a 7z file.

> Oh, I know the problem... I'm still using free on virtualalloc memory lol

Cool.  I'd never know.  I'll stop and re-start.  Current build is always at
https://ci.appveyor.com/project/MSP-Greg/ruby-loco

**Comment by ioquatix** [comment]
Any backtrace you provide is super useful to me since I'm not actually working on Windows and relying on appveyor and documentation to get things working.

**Comment by ioquatix** [comment]
`malloc`/`free` are general purpose memory allocation. They are not designed for stack allocation. But they can work if you align them correctly.

On Windows, I think it's the cause of intermittent failure. Sometimes aligned, sometimes not.

Linux has `mmap`/`mprotect`/`munmap`. It's lower level than `malloc`/`free` and provides more control over memory allocation, e.g. readonly, no access, etc.

On a stack, the last (lowest by address) page should be marked read-only or something similar. this is called guard page. If you try to make too many function calls, eventually your stack hits this page and it causes page fault. It kill your program, instead of allowing your program to overwrite arbitrary memory (a security problem).

On windows, it's the same concept, but they use `VirtualAlloc`/`VirtualProtect`/`VirtualFree`. Nothing is really different conceptually. You still allocate some page-aligned memory, put a guard page at the end (bottom) and use that as the stack.

**Comment by MSP-Greg** [comment]
Thanks for the explanation.

The last two builds of ruby-loco passed, with the patch skipping the two tests.  I can't get the test in test_enum.rb to fail locally, but the test in test_enumerable.rb fails consistently.

We've got a long weekend, I may build a repo to allow one to use a pre-existing build and run test-all & test-spec (or parts thereof) a half dozen times.  That might at least determine whether there is an intermittent issue.

**Comment by ioquatix** [comment]
I tried to reproduce bug on my local windows bug but I only managed to reproduce once, I don’t know how to attach debugger to `make test-all`.

I will try test_enumerable. How do you run just that test?

**Comment by MSP-Greg** [comment]
From test folder, I use:
```
ruby --disable=gems runner.rb ruby/test_enumerator.rb
```
Ran a third build, three in a row green.

**Comment by MSP-Greg** [comment]
@ioquatix

This is sort-of passing on Appveyor.  The error in test-all is an unrelated error.  But, if test-all fails, test-spec doesn't run.  Thanks, Greg

**Comment by ioquatix** [comment]
All builds passing, I merged to trunk.

**Comment by ioquatix** [comment]
For reference, https://github.com/ruby/ruby/commit/07a324a0f6464f31765ee4bc5cfc23a99d426705 was the merge commit.

**Pull request closed without merging**